exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/AsyncPageReader.java

```diff
@@ -0,0 +1,409 @@
+        .pushContext("Column: ", this.parentColumnReader.schemaElement.getName())
+        .pushContext("File: ", this.fileName).build(logger);
+    throw ex;
+  }
+
+  private DrillBuf decompress(PageHeader pageHeader, DrillBuf compressedData) {
+    DrillBuf pageDataBuf = null;
+    Stopwatch timer = Stopwatch.createUnstarted();
+    long timeToRead;
+    int compressedSize = pageHeader.getCompressed_page_size();
+    int uncompressedSize = pageHeader.getUncompressed_page_size();
+    pageDataBuf = allocateTemporaryBuffer(uncompressedSize);
+    try {
+      timer.start();
+      CompressionCodecName codecName = parentColumnReader.columnChunkMetaData.getCodec();
+      ByteBuffer input = compressedData.nioBuffer(0, compressedSize);
+      ByteBuffer output = pageDataBuf.nioBuffer(0, uncompressedSize);
+      DecompressionHelper decompressionHelper = new DecompressionHelper(codecName);
+      decompressionHelper.decompress(input, compressedSize, output, uncompressedSize);
+      pageDataBuf.writerIndex(uncompressedSize);
+      timeToRead = timer.elapsed(TimeUnit.NANOSECONDS);
+      this.updateStats(pageHeader, "Decompress", 0, timeToRead, compressedSize, uncompressedSize);
+    } catch (IOException e) {
+      handleAndThrowException(e, "Error decompressing data.");
+    }
+    return pageDataBuf;
+  }
+
+  @Override protected void nextInternal() throws IOException {
+    ReadStatus readStatus = null;
+    try {
+      Stopwatch timer = Stopwatch.createStarted();
+      readStatus = asyncPageRead.get();
+      long timeBlocked = timer.elapsed(TimeUnit.NANOSECONDS);
+      stats.timeDiskScanWait.addAndGet(timeBlocked);
+      stats.timeDiskScan.addAndGet(readStatus.getDiskScanTime());
+      if (readStatus.isDictionaryPage) {
+        stats.numDictPageLoads.incrementAndGet();
+        stats.timeDictPageLoads.addAndGet(timeBlocked + readStatus.getDiskScanTime());
+      } else {
+        stats.numDataPageLoads.incrementAndGet();
+        stats.timeDataPageLoads.addAndGet(timeBlocked + readStatus.getDiskScanTime());
+      }
+      pageHeader = readStatus.getPageHeader();
+      // reset this. At the time of calling close, if this is not null then a pending asyncPageRead needs to be consumed
+      asyncPageRead = null;
+    } catch (Exception e) {
+      handleAndThrowException(e, "Error reading page data.");
+    }
+
+    // TODO - figure out if we need multiple dictionary pages, I believe it may be limited to one
+    // I think we are clobbering parts of the dictionary if there can be multiple pages of dictionary
+
+    do {
+      if (pageHeader.getType() == PageType.DICTIONARY_PAGE) {
+        readDictionaryPageData(readStatus, parentColumnReader);
+        // Ugly. Use the Async task to make a synchronous read call.
+        readStatus = new AsyncPageReaderTask().call();
+        pageHeader = readStatus.getPageHeader();
+      }
+    } while (pageHeader.getType() == PageType.DICTIONARY_PAGE);
```

Comment:
Because this method has to block until it has the dictionary page(s) read. This part of the code is simply retaining the old code that seems to imply that more than one dictionary page may exist, though it is very unlikely to be so.